File: Python/Geek/linklist/link_list.py
from __future__ import annotations
from typing import List, Set, Dict, Tuple, Optional 
import sys


class Node(object):
    '''Node class to represent node of a linkList''' 


    def __init__(self,data:int,next:Node=None,prev:Node=None)->None:
        self.data = data
        self.next = next
        self.prev = prev  


    def get_data(self)-> int:
        return self.data 


    def set_data(self,data:int) -> None:
        self.data = data 


    def get_next(self)-> Node:
        return self.next 

    
    def set_next(self,n:Node) -> None:
        self.next = n 


    def get_prev(self)->Node:
        return self.prev  


    def set_prev(self,n:Node)-> None:
        self.prev = n 



class LinkList(object) :
    '''
    LinkList class 
    '''

    def __init__(self,head:Node=None)-> None:
        self.head = head 

    def set_head(self,head:Node) -> None:
        if head is not None:
            self.head = head
        else:
            print('head cannot be none') 


    def get_head(self) -> Node:
        return self.head 


    def create_link_list_from_list(self,ll:List[int])-> None:
        if ll is not None and len(ll) > 0:
            self.set_head(Node(ll[0])) 
            curr_node : Node = self.get_head()
            for element in ll[1:]:
                curr_node.set_next(Node(element))
                curr_node = curr_node.get_next()  

    def print_link_list(self) -> None:
        curr_node : Node = self.get_head()
        while(curr_node is not None):
            print(curr_node.get_data(), end = ' ')
            curr_node = curr_node.get_next()   


    def insert(self, data:int=None,prev_node_data:int=None)-> None:
        '''
        insert a node in LinkList after a given node .
        If no node is given then new node will be added at end of the linkList
        '''
        if prev_node_data is None:
            curr_node = self.get_head()
            curr_next = curr_node.get_next()
            while(curr_node is not None):
                if curr_next is not None:
                    curr_node = curr_next 
                    curr_next = curr_node.get_next() 
                else:
                    curr_node.set_next(Node(data))
                    break 
        else:
            curr_node = self.get_head() 
            curr_next = curr_node.get_next()
            while(curr_node is not None):
                if curr_node.get_data() != prev_node_data :
                    curr_node = curr_next
                    curr_next = curr_node.get_next() 
                else :
                    curr_node.set_next(Node(data))
                    curr_node.get_next().set_next(curr_next)  
                    break

            if curr_node is None:
                print('{0} is not present in LinkList'.format(prev_node_data)) 


    def delete_node(self,node_data:int) -> None:
        '''
        Delete node in LinkList
        '''
        if node_data is not None:
            curr_node : Node  = self.get_head()
            curr_prev : Node= None
            curr_next :Node = curr_node.get_next()
            while(curr_node is not None):
                if curr_node.get_data() != node_data :
                    curr_prev = curr_node
                    curr_node = curr_next 
                    curr_next = curr_node.get_next() 
                else:
                    if curr_node.get_data() != self.get_head().get_data() :
                        curr_prev.set_next(curr_next) 
                        break
                    else:
                        self.set_head(curr_next) 
                        break 


    def find_linklist_size_recursively(self)-> int:
        ''' 
        find size of link_list recursively
        '''
        if self.get_head() is not None:
            return self.__linklist_size(self.get_head(),0) 
        else:
            print('LinkList has no head')
            return 0

    
    def __linklist_size(self, node:Node,size:int) -> int:
        if node is not None:
            return self.__linklist_size(node.get_next() , 1 + size) 
        else:
            return size 


    def reverse_linklist_iteratively(self) -> None:
        '''
        reverse the link_list iteratively
        '''
        if self.get_head() is not None:
            curr_node : Node = self.get_head()
            curr_next : Node = None
            curr_prev : Node = None
            while curr_node is not None:
                curr_next = curr_node.get_next()
                curr_node.set_next(curr_prev)
                curr_prev = curr_node 
                curr_node = curr_next 
            self.set_head(curr_prev) 



    def reverse_linklist_recursively(self) -> None:
        '''
        reverse linklist recursively
        '''
        return self._reverse_linklist_recursively(self.get_head()) 


    def _reverse_linklist_recursively(self,cur_node:Node,cur_prev:Node=None) -> None:
        if cur_node is not None:
            cur_next:Node = cur_node.get_next() 
            cur_node.set_next(cur_prev) 
            cur_prev = cur_node 
            cur_node = cur_next  
            return self._reverse_linklist_recursively(cur_node,cur_prev=cur_prev)

        self.set_head(cur_prev) 


    def reverseK(self, k:int) -> None:
        '''
        reverse LinkList in  group of K
        '''
        self._reverseK(self.get_head(),k) 

    
    def _reverseK(self,node:Node, k:int) -> Node: 
        cur_node:Node = node
        cur_prev:Node = None
        cur_next:Node = None
        count:int = 0 
        while cur_node is not None and count < k: 
            cur_next = cur_node.get_next()
            cur_node.set_next(cur_prev)
            cur_prev = cur_node
            cur_node = cur_next 
            count = count + 1 
        if cur_next is not None:
            self.get_head().set_next(self._reverseK(cur_next,k))

        self.set_head(cur_prev)
        return cur_prev





    
    











    

File: Java/topDS/src/main/java/datastructure/linkedlist/LinkList.java
package datastructure.linkedlist;

import java.util.List;

import datastructure.graph.CriticalRouterNetwork;

public class LinkList<T extends Comparable<T>> { 
	
	
	public Node<T> head;
	

	
	void setHead(Node<T> head) {
		if(head !=null) {
			this.head = head;
			}
	}
	
	
	// Add new node at start of the linkList and create a doubly connection
	void push(T data) {
		
		//if no head is set
		if(this.head == null) {
			  this.head = new Node<>(data);
		 }
		else {
		Node<T> node = new Node<>(data);
		node.next = this.head;
		this.head.previous = node;
		this.setHead(node);
		}
	 }
	
	
	 /**
	   * Add a node next to given node (cannot add new node at the start of linkedList)
	   *
	   * @param newNodeData Data value of new node
	   * @param insertAfterNodeData Data value of given node
	   */
	  public void insertAfter(T newNodeData, T insertAfterNodeData) {
	    // create the new_node
	    Node<T> new_node = new Node<T>(newNodeData);

	    // search the existing node with given data value starting at head of the
	    // LinkedList
	    Node<T> curr = this.head;
	    Node<T> currNext = curr.next;

	    // iterate over the linkedList till the node is found or end of the LinkedList
	    // is reached
	    while (curr != null) {
	      if (curr.data.equals(insertAfterNodeData)) {
	        break;
	      }

	      curr = currNext;
	      currNext = curr.next;
	    }
	    if (curr == null) {
	      System.out.println("LinkedList has no Node with the value :" + insertAfterNodeData);
	    } else {
	      curr.setNext(new_node);
	      new_node.previous = curr;
	      new_node.setNext(currNext);
	      currNext.setPrevious(new_node);
	    }
	  } 
	  
	  
	  // Delete a node in LinkList with given value
	  public void deleteNode(T data) {
		  Node<T> curr = this.head;
		  Node<T> currPrev = this.head;
		  
		  while(curr !=null) {
			  if(curr.data.compareTo(data)==0) break;
			  curr = curr.next;
			  currPrev = curr;
		  }
		  
		  if(curr.data.compareTo(this.head.data)==0) {
			  this.setHead(curr.next);
			  curr.next.previous = currPrev;
		  }
		  
		  //Assuming no previous connection
		  else if(curr.data.compareTo(this.head.data)==0 && curr !=null) {
			  currPrev.next = curr.next;
			  curr.next.previous = currPrev;
		  }
		  
		  else {
			  System.out.println("Node with given data value is not found");
		  }
		  
	  }
	  
	  //Find size of LinkList recursively
	  public int LinkListSize(Node<T> node) {
		  if(node == null) return 0;
		  else return 1 + LinkListSize(node.next);
		  
	  }
	  
	  
	  //reverse LinkList recursively
	  public void reverseList(Node<T> curr, Node<T> currNext,Node<T> currPrev) {
		  if(currNext != null) {
			  
			  Node<T> node = currNext.next;
			  
			  //change the links
			  curr.next = currPrev;
			  currNext.next = curr;
			  
			  //move one farther
			  currPrev = curr;
			  curr = currNext;
			  currNext = node;
			  
			  this.setHead(curr);
			  this.reverseList(curr, currNext,currPrev);
		  }
	  }
	  
	  
	  //reverse LinkList recursively
	  /**
	   * REMEMBER Reverse a LinkList in group of K
	   *
	   * @param head
	   * @param k
	   * @return
	   */
	  public Node<T> reverseListKGroup(Node<T> head, int k) {
	
		Node<T> prev = null;
	    Node<T> curr = head;
	    Node<T> next = null;
	    

	    int count = 0;

	    while (count < k && curr != null) {
	      next = curr.next;
	      curr.next = prev;
	      prev = curr;
	      curr = next;
	      count++;
	    }

	    /**
	     * Next is now a pointer to (K+1)th node Recursively call the list, starting from next. And make
	     * rest of the list as next of first node
	     */
	    if (curr != null) {
	    	head.next = reverseListKGroup(curr, k);

	    }
	    

	    // prev is now head of the input list
	    return prev;
	  }
	
	  /**
	   * Create LinkList from List
	   * @author krish.mahajan@ibm.com
	   *
	   * @param <T>
	   */
	  public void createLinkListFromCollection(List<T> list) {
		  for(T data : list) {  
		  this.push(data);
		  }
	  }
	  
	  //Print LinkList
	  public void printLinkList() {
		  if(this.head != null) {
			  Node<T> curr = this.head;
			  while(curr!=null) {
				  System.out.print(curr.data + " ");
				  curr = curr.next;
			  }
		  }
	  }
	  
	  
	  
	  
	
	
	
	static class Node<T extends Comparable<T>>{
		
		private T data;
		public Node next;
		public Node previous;
		
		Node(T data) {
			this.data = data;
		}
		
		void setNext(Node node) {
			this.next = node;
		}
		
		void setPrevious(Node node) {
			this.previous = node;
		}
	}
}
